cvat-ui/plugins/sam/src/ts/index.tsx

```diff
@@ -183,106 +213,135 @@ const samPlugin: SAMPlugin = {
-                        return toMatImage(input, width, height);
-                    }
-
-                    const data = await (plugin.data.session as InferenceSession).run(feeds);
-                    const { masks, low_res_masks: lowResMasks } = data;
-                    const imageData = onnxToImage(masks.data, masks.dims[3], masks.dims[2]);
-                    plugin.data.lowResMasks.set(key, lowResMasks);
-
-                    const xtl = Number(data.xtl.data[0]);
-                    const xbr = Number(data.xbr.data[0]);
-                    const ytl = Number(data.ytl.data[0]);
-                    const ybr = Number(data.ybr.data[0]);
-
-                    return {
-                        mask: imageData,
-                        bounds: [xtl, ytl, xbr, ybr],
-                    };
+                        plugin.data.jobs = {
+                            // we do not need to store old job instances
+                            [job.id]: job,
+                        };
+
+                        job.frames.get(frame)
+                            .then(({ height: imHeight, width: imWidth }: { height: number; width: number }) => {
+                                const key = `${taskID}_${frame}`;
+
+                                if (result) {
+                                    const bin = window.atob(result.blob);
+                                    const uint8Array = new Uint8Array(bin.length);
+                                    for (let i = 0; i < bin.length; i++) {
+                                        uint8Array[i] = bin.charCodeAt(i);
+                                    }
+                                    const float32Arr = new Float32Array(uint8Array.buffer);
+                                    plugin.data.embeddings.set(key, new Tensor('float32', float32Arr, [1, 256, 64, 64]));
+                                }
+
+                                const modelScale = {
+                                    width: imWidth,
+                                    height: imHeight,
+                                    samScale: getModelScale(imWidth, imHeight),
+                                };
+
+                                const composedClicks = [...pos_points, ...neg_points].map(([x, y], index) => ({
+                                    clickType: index < pos_points.length ? 1 : 0 as 0 | 1,
+                                    height: null,
+                                    width: null,
+                                    x,
+                                    y,
+                                }));
+
+                                const feeds = modelData({
+                                    clicks: composedClicks,
+                                    tensor: plugin.data.embeddings.get(key) as Tensor,
+                                    modelScale,
+                                    maskInput: plugin.data.lowResMasks.has(key) ?
```

Comment:
Also not the best solution, please look at efficientvit

Reply:
It is buggy when removing and optionally adding new points.

Reply:
Added logic with `lastClicks`